src/System.Linq/tests/ToArrayTests.cs

```diff
@@ -199,5 +199,95 @@ public void ToArray_ListWhereSelect(int[] sourceIntegers, string[] convertedStri
             Assert.Equal(Array.Empty<string>(), sourceList.Select(i => i.ToString()).Where(s => s == null).ToArray());
         }
 
+        [Fact]
+        public void SameResultsRepeatCallsFromWhereOnIntQuery()
+        {
+            var q = from x in new[] { 9999, 0, 888, -1, 66, -777, 1, 2, -12345 }
+                    where x > Int32.MinValue
+                    select x;
+
+            Assert.Equal(q.ToArray(), q.ToArray());
+        }
+        
+        [Fact]
+        public void SameResultsRepeatCallsFromWhereOnStringQuery()
+        {
+            var q = from x in new[] { "!@#$%^", "C", "AAA", "", "Calling Twice", "SoS", String.Empty }
+                        where !String.IsNullOrEmpty(x)
+                        select x;
+
+            Assert.Equal(q.ToArray(), q.ToArray());
+        }
+
+        [Fact]
+        public void SourceIsEmptyICollectionT()
+        {
+            int[] source = { };
+            int[] expected = { };
+
+            ICollection<int> collection = source as ICollection<int>;
+
+            Assert.Equal(expected, source.ToArray());
+            Assert.Equal(expected, collection.ToArray());
+        }
+
+        [Fact]
+        public void SourceIsICollectionTWithFewElements()
+        {
+            int?[] source = { -5, null, 0, 10, 3, -1, null, 4, 9 };
+            int?[] expected = { -5, null, 0, 10, 3, -1, null, 4, 9 };
+
+            ICollection<int?> collection = source as ICollection<int?>;
```

Comment:
That it tests on both `source` and `collection` was my addition; the original checked only `source` after confirming it was indeed an `ICollection<T>`, so checking that both the collection-based and non-collection-based tests worked. I figured since there was another variable there, there was no harm checking that some future change didn't do anything weird based on the type.